# CROP-seq
Repository for the processing and analysis of CROP-seq data. This repository contains code to do the following steps for CROP-seq analysis:

1. Reference preparation
2. Guide quantification from scRNA-seq data
3. Guide quantificationf from enriched RNAseq data

## Reference preparation
The reference requires the following information:

- Primary assembly of the organism ([FASTA](ftp://ftp.ensembl.org/pub/release-98/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz) and [GTF](ftp://ftp.ensembl.org/pub/release-98/gtf/homo_sapiens/Homo_sapiens.GRCh38.98.chr.gtf.gz))
- [Spreadsheet with gRNA sequence and target genes](example_data/GWAS_sgRNAs.csv)
- [Spreadsheet with CROP-seq vector elements](‎⁨example_data/CROP-seq.csv⁩)

1. Use the following script: [buildReference.py](py/buildReference.py) to build the guide reference.

```bash
python buildReference.py -i example_data/GWAS_sgRNAs.csv -c example_data/CROP\-seq.csv -o Test
```

2. Concatenate the generated files to your primary assembly files, to create a new set of reference files.

```bash
> cat Homo_sapiens.GRCh38.dna.primary_assembly.fa CROPseq.fa > Homo_sapiens.CROPseq.fa
> cat Homo_sapiens.GRCh38.98.chr.gtf CROPseq.gtf > Homo_sapiens.CROPseq.gtf
```

3. Convert the files to Cell Ranger reference files using *cellranger mkref*.

```bash
# Path to input files
GENOME_NAME=Homo_sapiens.CROPseq
INPUT_FASTA=Homo_sapiens.CROPseq.fa
INPUT_GTF=Homo_sapiens.CROPseq.gtf
INPUT_FILTERED_GTF=Homo_sapiens.CROPseq.Filtered.gtf

# Filter GTF
cellranger mkgtf ${INPUT_GTF} ${INPUT_FILTERED_GTF}  --attribute=gene_biotype:protein_coding --attribute=gene_biotype:lincRNA --attribute=gene_biotype:antisense || exit 1

# Create STAR reference
cellranger mkref --genome=${GENOME_NAME} --fasta=${INPUT_FASTA} --genes=${INPUT_FILTERED_GTF} || exit 1
```

## scRNA-seq Processing and Analysis
###  Chromium Processing
This step is usually done by the service provider. Samples need to be mapped to the reference prepared in the previous step.

### Guide molecule identification
If the reference has worked, you will receive a count table with the guide and control RNA as additional genes in the expression matrix. These can be retrieved using the following steps.

## MiSeq Enrichment
### FASTQ Preparation
This uses merged paired-end reads as generated by the Galaxy pipeline.

1. Label reads with combined cell barcode and UMI sequence, located at the first 28 nucleotides of the merged sequences. This can be done with [fastp](https://github.com/OpenGene/fastp). 

```bash
fastp -i Galaxy112-[ARRAY1_R1_R2_merge].fastqsanger  -o ARRAY1.fastq -U --umi_loc read1 --umi_len 28  -w 4
```

2. Strip polyT tail from the reads using [cutadapt](https://cutadapt.readthedocs.io/en/stable/index.html).

```bash
cutadapt -j 4 -g "T{200}" -o ARRAY1_Processed.fastq ARRAY1.fastq 
```

This can be run for all samples using the preprocessFastq.sh script.

## MiSeq Enrichment
### Mapping
Map guides to the reference generated in the first section using STAR aligner.

```bash
STAR --genomeDir $REF_DIR --readFilesCommand cat \
--readFilesIn $INPUT_FILE \
--outSAMtype BAM Unsorted \
--quantMode GeneCounts \
--runThreadN 16 \
--outFileNamePrefix $SAMPLE_NAME \
--outReadsUnmapped Fastx \
--outFilterScoreMinOverLread 0 \
--outFilterMatchNminOverLread 0 \
--outFilterMatchNmin 0 \
--outFilterMismatchNmax 2
```

This will output a BAM file and a count table that we will use to trace back to the original cells.

### Quantification via reads
Run [quantifyEnrichment.py](py/quantifyEnrichment.py).
